packages/compass-editor/src/codemirror/query-autocompleter.test.ts

```diff
@@ -0,0 +1,87 @@
+import { EditorView } from '@codemirror/view';
+import { EditorState, EditorSelection } from '@codemirror/state';
+import type {
+  CompletionSource,
+  CompletionResult,
+} from '@codemirror/autocomplete';
+import { CompletionContext } from '@codemirror/autocomplete';
+import { expect } from 'chai';
+// import { languages } from '../json-editor';
+import { createQueryAutocompleter } from './query-autocompleter';
+
+const setupCompleter = <T extends (...args: any[]) => CompletionSource>(
+  completer: T
+) => {
+  const el = window.document.createElement('div');
+  window.document.body.appendChild(el);
+  const editor = new EditorView({
+    state: EditorState.create({
+      doc: '',
+    }),
+    selection: EditorSelection.single(0),
+    // TODO: doesn't look like extensions work when inside jsdom. need to sort
+    // it out to be able to better test completers
+    // extensions: [languages.javascript()],
+    parent: el,
+  });
+  const getCompletions = (text = '', ...args: Parameters<T>) => {
+    editor.dispatch({
+      changes: { from: 0, to: editor.state.doc.length, insert: text },
+      selection: { anchor: text.length },
+      userEvent: 'input.type',
+    });
+    return (
+      (
+        completer(...args)(
+          new CompletionContext(editor.state, text.length, false)
+        ) as CompletionResult
+      )?.options ?? []
+    );
+  };
+  const cleanup = () => {
+    editor.destroy();
+    el.remove();
+  };
+  return { getCompletions, cleanup };
+};
+
+describe('query autocompleter', function () {
+  const { getCompletions, cleanup } = setupCompleter(createQueryAutocompleter);
+
+  after(cleanup);
+
+  it('returns all completions when current token is vaguely matches identifier', function () {
+    expect(getCompletions('foo')).to.have.lengthOf(45);
+  });
+
+  it("doesn't return anything when not matching identifier", function () {
+    expect(getCompletions('[')).to.have.lengthOf(0);
+  });
+
+  // See above, need extensions to actually work for this
+  it.skip('completes "any text" when inside a string', function () {
+    expect(
+      getCompletions('{ bar: 1, buz: 2, foo: "b').map(
+        (completion) => completion.label
+      )
+    ).to.deep.eq(2);
+  });
+
+  it('escapes field names that are not valid identifiers', function () {
```

Comment:
Do we have a test for the autocompleting contents without escaped field names?

Reply:
I have a test that checks that autocompleter returns 45 results when cursor is at something that passes as identifier [here](https://github.com/mongodb-js/compass/pull/4134/files#diff-fe7dc1a3bb86e69ba68782f28542ad8fb203eced35998c42d41bbd0d614a0aa0R53-R55). Codemirror doesn't work exactly the way ace works, so in case of codemirror we can avoid filtering of the autocompletion items ourselves, we return everything and codemirror fuzzy matchers filters them. Do you mean something else? Would you prefer we do exact match on values from compass-constants package?

Reply:
Ah yup. All good